apps/web/src/shared/gql.ts

```diff
@@ -0,0 +1,105 @@
+import { prefixUrl } from '@affine/env/api';
+import type {
+  GraphQLQuery,
+  MutationOptions,
+  QueryOptions,
+  QueryResponse,
+  QueryVariables,
+} from '@affine/graphql';
+import { gqlFetcherFactory } from '@affine/graphql';
+import type { GraphQLError } from 'graphql';
+import type { SWRConfiguration, SWRResponse } from 'swr';
+import useSWR from 'swr';
+import type {
+  SWRMutationConfiguration,
+  SWRMutationResponse,
+} from 'swr/mutation';
+import useSWRMutation from 'swr/mutation';
+
+const fetcher = gqlFetcherFactory(prefixUrl + '/graphql');
+
+/**
+ * A `useSWR` wrapper for sending graphql queries
+ *
+ * @example
+ *
+ * ```ts
+ * import { someQuery, someQueryWithNoVars } from '@affine/graphql'
+ *
+ * const swrResponse1 = useQuery({
+ *   query: workspaceByIdQuery,
+ *   variables: { id: '1' }
+ * })
+ *
+ * const swrResponse2 = useQuery({
+ *   query: someQueryWithNoVars
+ * })
+ * ```
+ */
+export function useQuery<Query extends GraphQLQuery>(
+  options: QueryOptions<Query>
+): SWRResponse<QueryResponse<Query>, GraphQLError | GraphQLError[]>;
+export function useQuery<Query extends GraphQLQuery>(
+  options: QueryOptions<Query>,
+  config: Omit<
+    SWRConfiguration<
+      QueryResponse<Query>,
+      GraphQLError | GraphQLError[],
+      typeof fetcher<Query>
+    >,
+    'fetcher'
+  >
+): SWRResponse<QueryResponse<Query>, GraphQLError | GraphQLError[]>;
+export function useQuery(options: QueryOptions<any>, config?: any) {
+  return useSWR(options.query.id, () => fetcher(options), config);
```

Comment:
You might also want to include `options.variables` as the `key` here.

```tsx
const swrResponse1 = useQuery({
 query: workspaceByIdQuery,
 variables: { id: '1' }
})
const swrResponse2 = useQuery({
 query: workspaceByIdQuery,
 variables: { id: '2' }
})
```
Then `swrResponse1` and `swrResponse2` could be cached individually